### DISCUSSION TITLE
SetDefault is setting IsSet

### DISCUSSION THREAD

**Discussion by itaysk** [spf13/viper#1766] [Q&A]
I'm not sure this is by design or a bug, but I found out that `SetDefault` is making the `flag IsSet` which I didn't expect it to, since the user had never set the flag. Giving the flag a default value during construction does not make it Set, but Giving it a default afterwards does.
Example:

```go
package main

import (
	"fmt"

	"github.com/spf13/pflag"
	"github.com/spf13/viper"
)

func main() {
	pflag.String("test", "testing", "")
	tFlag := pflag.Lookup("test")
	viper.BindPFlag("test", tFlag)

	fmt.Println(viper.IsSet("test"))
	fmt.Println(viper.GetString("test"))
	viper.SetDefault("test", "testing123")
	fmt.Println(viper.IsSet("test"))
	fmt.Println(viper.GetString("test"))
}
```

```text
$ go run main.go
false
testing
true
testing123
```

**Comment by hexvalid** [comment]
This is the distinction between Viper's value lookup and pflag's tracking of explicitly supplied flags.

`viper.IsSet("test")` checks whether Viper can resolve a value through its normal sources, including values registered with `SetDefault`. An unchanged pflag's construction default is treated differently: it is a fallback for `Get`, but is excluded from the lookup used by `IsSet`. That accounts for all four lines in your example.

To determine whether the user supplied the CLI flag, inspect pflag after parsing:

```go
pflag.Parse()
explicitlyProvided := pflag.Lookup("test").Changed
```

Calling `viper.SetDefault` does not change that `Changed` field. I checked your sequence with Viper v1.21.0: `IsSet` changes from false to true, while `Changed` stays false; parsing `--test=explicit` makes `Changed` true.

For presence specifically in a configuration file, use `viper.InConfig("test")`. These answer different questions from whether an effective value is available.

Sources: [Viper's `IsSet` / `find`](https://github.com/spf13/viper/blob/v1.21.0/viper.go) and [pflag.Flag.Changed](https://pkg.go.dev/github.com/spf13/pflag#Flag).
